### PR TITLE
restmapper: re-try shortcut expander after not-found error

### LINKED ISSUE #65517: Getting new CRD with short name/category will occur error
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>

/kind bug

> /kind feature

**What happened**:

When I created a crd e.g.:

```yam
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  # name must match the spec fields below, and be in the form: <plural>.<group>
  name: crontabs.stable.example.com
spec:
  # group name to use for REST API: /apis/<group>/<version>
  group: stable.example.com
  # version name to use for REST API: /apis/<group>/<version>
  version: v1
  # either Namespaced or Cluster
  scope: Namespaced
  names:
    # plural name to be used in the URL: /apis/<group>/<version>/<plural>
    plural: crontabs
    # singular name to be used as an alias on the CLI and for display
    singular: crontab
    # kind is normally the CamelCased singular type. Your resource manifests use this.
    kind: CronTab
    # shortNames allow shorter string to match your resource on the CLI
    shortNames:
    - ct
    categories:
    - testcat
```
Then I used `kubectl get testcat`. I found that the first time the result will be with an error:

> error: the server doesn't have a resource type "testcat"

And as for the second time it will be normal:

> No resources found. 

And it can be stably reproduced every time I created a new CRD.

**What you expected to happen**:

kubectl returns `No resources found. ` the first time I requested the new CRD instead of an error.

**Anything else we need to know?**:

I also found that requesting short name e.g. `kubectl get ct` will be the same. It returns an error the first time but becomes normal the second time.😲

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/kind bug

### PR DESCRIPTION
**Pull request by soltysh**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Currently first kubectl invocation using short name ends with a not found error. This is because the shortcut expander is run before the discovery is refreshed. The proposed approach in this PR is to re-try shortcut expansion after a not found error coming from discovery. 

This should not increase apiserver requests, since already after initial `KindFor` see: https://github.com/openshift/kubernetes/blob/01aa0f3f6052c9866dcb69054d8924fa8c96e3cb/staging/src/k8s.io/client-go/restmapper/discovery.go#L225-L236 we invalidate cache and re-fetch data from the server. So the duplicated call will built on top of that data. 

**Which issue(s) this PR fixes**:
Fixes #65517

**Special notes for your reviewer**:
I'm putting this on
/hold
to verify options and gather feedback.

/assign @caesarxuchao 

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/restmapper/shortcut.go [hunk 1]
```diff
@@ -43,7 +43,11 @@ func NewShortcutExpander(delegate meta.RESTMapper, client discovery.DiscoveryInt
 
 // KindFor fulfills meta.RESTMapper
 func (e shortcutExpander) KindFor(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
-	return e.RESTMapper.KindFor(e.expandResourceShortcut(resource))
+	gvk, err := e.RESTMapper.KindFor(e.expandResourceShortcut(resource))
+	if meta.IsNoMatchError(err) {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/restmapper/shortcut.go:47]: would have expected to call `e.expandResourceShortcut(resource)` first and rerun on error or empty result. Isn't `e.expandResourceShortcut` failing?

Comment (soltysh) [reply staging/src/k8s.io/client-go/restmapper/shortcut.go:47]: Currently, `expandResourceShortcut` will just return resource as is if it doesn't find shortcut, but yeah I could return error or empty and based on that re-run expansion. lemme fix that...

**Comment by dims** [comment]
i chased a rabbit down a hole [1] and ended up here. Hope we can fix https://github.com/kubernetes/kubernetes/issues/65517 in 2021 :)

[1] https://testgrid.k8s.io/sig-node-containerd#e2e-cos-flaky&exclude-non-failed-tests=

**Comment by soltysh** [comment]
Still a thing
/reopen

**Comment by xmcqueen** [comment]
@dims That's exactly how I got here too

@soltysh is there no consensus on this or has it not been reviewed?  It looks like there's a comment awaiting an update here:

https://github.com/kubernetes/kubernetes/pull/96771#discussion_r528849193

File: staging/src/k8s.io/client-go/restmapper/shortcut.go (hunk 1 above)
Comment (xmcqueen) [reply staging/src/k8s.io/client-go/restmapper/shortcut.go:47]: @soltysh hi, thanks for getting this PR started.  There's some interest in getting this going.  We're trying to fix a large number flaky tests.  Can you get the above patch going?

**Comment by dims** [comment]
@xmcqueen try internal RH channels? :) (both Maciej and Stefan) unfortunately for all of us github notifications are pretty much unusable.

**Comment by soltysh** [comment]
@xmcqueen @dims sorry for the delay, I wanted to add a proper unit test to ensure this behavior works as expected. I've already pinged @sttts for a review.

**Comment by sttts** [comment]
The the comment lgtm.

/lgtm
/approve

**Comment by xmcqueen** [comment]
thanks @soltysh !

**Pull request merged**